include/range/v3/algorithm/for_each_n.hpp

```diff
@@ -0,0 +1,67 @@
+/// \file
+// Range v3 library
+//
+//  Copyright Eric Niebler 2014
+//  Copyright Rostislav Khlebnikov 2017
+//
+//  Use, modification and distribution is subject to the
+//  Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at
+//  http://www.boost.org/LICENSE_1_0.txt)
+//
+// Project home: https://github.com/ericniebler/range-v3
+//
+#ifndef RANGES_V3_ALGORITHM_FOR_EACH_N_HPP
+#define RANGES_V3_ALGORITHM_FOR_EACH_N_HPP
+
+#include <functional>
+#include <range/v3/range_fwd.hpp>
+#include <range/v3/range_traits.hpp>
+#include <range/v3/range_concepts.hpp>
+#include <range/v3/algorithm/tagspec.hpp>
+#include <range/v3/utility/functional.hpp>
+#include <range/v3/utility/iterator_traits.hpp>
+#include <range/v3/utility/static_const.hpp>
+#include <range/v3/utility/tagged_pair.hpp>
+
+namespace ranges
+{
+    inline namespace v3
+    {
+        /// \addtogroup group-algorithms
+        /// @{
+        struct for_each_n_fn
+        {
+            template<typename I, typename F, typename P = ident,
+                CONCEPT_REQUIRES_(InputIterator<I>() &&
+                    IndirectInvocable<F, projected<I, P>>())>
+            tagged_pair<tag::in(I), tag::fun(F)>
```

Comment:
Please bear with me - just trying to fully understand the logic behind the design decisions so there's less time required from you for reviewing future pull requests from me :)

Here's how I understand it:

- STL `for_each` doesn't return the iterator because it takes two iterators of the same type as `first` and `last`, so the user has the access to `last` anyway and the algorithm wouldn't provide additional info by returning the final iterator
- Range `for_each` may take a sentinel, so the final iterator might contain some more information than the user has. But because STL returns the function object - range `for_each` returns both with a tagged pair. But because the STL `for_each_n` returns the iterator - so does range `for_each_n` and there's actually no need to return the function object (for reasons you mentioned).

Any misunderstandings from my side?